content/blog/entries/crawling-500-million/contents.lr

```diff
@@ -0,0 +1,485 @@
+
+1. It may not be correct to assume that performance will steadily degrade instead of failing all at once.
+2. We can't detect whether we are the cause of a performance issue or if they're simply experiencing server trouble due to configuration errors or high traffic. We could get stuck in local maximums due to normal fluctuations in traffic.
+3. Recording TTFB in Python is difficult because it requires low level access to connection data. We might have to write an extension to `aiohttp` to get it.
+
+Eventually I decided that this is too much hassle. Can we get the job done with a simpler strategy?
+
+It turns out that the size of a website is typically correlated with infrastructure capabilities. The reasoning behind this is that if you are capable of hosting 450MM images, you are probably able to handle at least a couple hundred requests per second for serving traffic. In our case, we already know how many images a source has, so it's easy for us to peg our rate limit between a low minimum for small websites and a reasonable maximum for large websites, and then interpolate the rate limit for everything in between.
+
+Of course, it's important to note that this is only a rough heuristic that we use to make a reasonable guess about what a website can handle. We have to allow the possibility that we set our rate limit too aggressively in spite of our precautions.
+
+#### Backing off with circuit breakers
+If our heuristic fails to correctly approximate the bandwidth capabilities of a site, we are going to start encountering problems. For one, we might exceed the server-side rate limit, which means we will see `429 Rate Limit Exceeded` and `403 Forbidden` errors instead of the images we're trying to crawl. Worse yet, the upstream source might continue to happily serve requests while we suck up all of their traffic capacity, resulting in other users being unable to view the images. Clearly, in either scenario, we need to either reduce our crawl rate or even give up crawling the source entirely if it appears that we are impacting their uptime.
+
+To handle these situations, we have two tools in our toolbox: a sliding window recording the status code of every request made we've made to each domain in the last 60 seconds, and a list of the last 50 status for each website. If the number of errors in our one minute window exceed 10%, something is wrong; wait a minute before trying again. If we have encountered many errors in a row, however, that suggests that we're having trouble with a particular site, so we ought to give up crawling the source and raise an alert.
```

Comment:
"last 50 status for each website" is unclear.

Reply:
This sentence was a little hard to parse for me: "If the number of errors in our one minute window exceed 10%, something is wrong; wait a minute before trying again." Perhaps something like "we should wait a minute before trying again"?